### ISSUE TITLE
State subclasses

### ISSUE THREAD

**Issue by cliffckerr** [starsimhub/starsim#407]
Instead of a single `ss.State()`, I think it would be clearer to have subclasses that were specific for each  type, e.g. bool, int, float, since it would be an error to substitute one for the other (e.g. define `susceptible` as a float). I also personally feel that `State` makes sense only for bool arrays (`alive` is a state, but `rel_sus` to me is more of a property/parameter). So I suggest that instead of (from `disease.py`):
```py
ss.State('susceptible', bool, True),
ss.State('infected', bool, False),
ss.State('rel_sus', float, 1.0),
ss.State('rel_sev', float, 1.0),
ss.State('rel_trans', float, 1.0),
ss.State('ti_infected', int, ss.INT_NAN),
```
we have
```py
ss.BoolArr('susceptible', True), # no default value for boolean arrays
ss.BoolArr('infected', False),
ss.FloatArr('rel_sus', 1.0), # np.nan would be the default fill value
ss.FloatArr('rel_sev', 1.0),
ss.FloatArr('rel_trans', 1.0),
ss.IntArr('ti_infected'), # ss.INT_NAN would be the default fill value so doesn't need to be specified
```

If users want other dtypes, they could use `ss.State` directly (which I would probably suggest renaming `ss.Arr`).

This would also have the advantage that we could optimize computations. For example, if we use `np.int32` as the dtype for `IntArr` (which I propose we should!), and we use `np.int32()` on the inputs, then we should get a 2x speedup on the ufunc arithmetic. We could still do this from within a single `State` class, but I think it would be less clear, and then the onus would be on the user to remember to do `ss.State('ti_infected', np.int32, ss.INT_NAN)` everywhere, which is getting ugly.

**Comment by daniel-klein** [comment]
This seems like a good place to use options in settings.py. We can autodetect the data type consistently, so it seems to be a question of precision, right?

**Comment by cliffckerr** [comment]
Closed by #456
